ksql-engine/src/main/java/io/confluent/ksql/util/ExpressionTypeManager.java

```diff
@@ -89,251 +97,337 @@ public void setSchema(final Schema schema) {
+        final LogicalBinaryExpression node,
+        final ExpressionTypeContext context) {
+      process(node.getLeft(), context);
+      process(node.getRight(), context);
       return null;
     }
-    if (node.getName().getSuffix().equalsIgnoreCase(FetchFieldFromStruct.FUNCTION_NAME)) {
-      process(node.getArguments().get(0), expressionTypeContext);
-      final Schema firstArgSchema = expressionTypeContext.getSchema();
-      final String fieldName = ((StringLiteral) node.getArguments().get(1)).getValue();
-      if (firstArgSchema.field(fieldName) == null) {
-        throw new KsqlException(String.format("Could not find field %s in %s.",
-            fieldName,
-            node.getArguments().get(0).toString()));
-      }
-      final Schema returnSchema = firstArgSchema.field(fieldName).schema();
-      expressionTypeContext.setSchema(returnSchema);
-    } else {
-      final UdfFactory udfFactory = functionRegistry.getUdfFactory(node.getName().getSuffix());
-      final List<Schema> argTypes = new ArrayList<>();
-      for (final Expression expression : node.getArguments()) {
-        process(expression, expressionTypeContext);
-        argTypes.add(expressionTypeContext.getSchema());
-      }
-      final Schema returnType = udfFactory.getFunction(argTypes)
-          .getReturnType(argTypes);
-      expressionTypeContext.setSchema(returnType);
+
+    @Override
+    public Void visitType(
+        final Type type,
+        final ExpressionTypeContext expressionTypeContext
+    ) {
+      throw VisitorUtil.illegalState(this, type);
     }
-    return null;
-  }
 
-  private static Schema resolveArithmeticType(
-      final Schema leftSchema,
-      final Schema rightSchema,
-      final Operator operator) {
-    return SchemaUtil.resolveBinaryOperatorResultType(leftSchema, rightSchema, operator);
-  }
+    @Override
+    public Void visitTimeLiteral(
+        final TimeLiteral timeLiteral,
+        final ExpressionTypeContext expressionTypeContext
+    ) {
+      throw VisitorUtil.unsupportedOperation(this, timeLiteral);
+    }
 
-  private void validateSearchedCaseExpression(final SearchedCaseExpression searchedCaseExpression) {
-    final Schema firstResultSchema = getExpressionSchema(
-        searchedCaseExpression.getWhenClauses().get(0).getResult());
-    searchedCaseExpression.getWhenClauses()
-        .forEach(whenClause -> validateWhenClause(whenClause, firstResultSchema));
-    searchedCaseExpression.getDefaultValue()
-        .map(this::getExpressionSchema)
-        .filter(defaultSchema -> !firstResultSchema.equals(defaultSchema))
-        .ifPresent(badSchema -> {
-          throw new KsqlException("Invalid Case expression."
-              + " Schema for the default clause should be the same as schema for THEN clauses."
-              + " Result scheme: " + firstResultSchema + "."
-              + " Schema for default expression is " + badSchema);
-        });
-  }
+    @Override
```

Comment:
The diff for this file looks big, but its really just the functions below, plus the visitor part has been moved to an inner class.